### PR TITLE
enroll host into IPA doman in domain-level 0 replica install

### PR DESCRIPTION
**Pull request by martbab**
This PR implements host enrollment during installation of domain level 0
replica so that it will be possible to merge both domain levels into one
workflow.

Couple of discussion points:
    \* should we use preserve_enrolled_state decorator to automatically unenroll
      client from IPA domain? this was not done before so the user had to clean
      the host entry, etc. himself if something exploded

```
* should this be in install or install_check? I am ok with install_check
  but will need access to connected remote-api to create host entry first

* should we add the host to ipaservers group right away or reuse logic from
  promote code to add it later?
```

### CODE REVIEW

**Comment by stlaz** [comment]
1. Should it work it would be lovely.
2. Either do it in install and let install_check perform all checks prior to it so that you don't have to be bothered by removing it or do it at the end of install_check (will have to move it to promote_check as well).
3. I would keep the same logic as is in promote.

File: ipalib/install/host.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                # 3 - Unable to open keytab
+                # 5 - Principal name or realm not found in keytab
+                root_logger.error("Error trying to clean keytab: " +
+                    "/usr/sbin/ipa-rmkeytab returned %s" % e.returncode)
+        else:
+            root_logger.info("Removed old keys for realm %s from %s" % (
+                realm, paths.KRB5_KEYTAB))
+
+
+def configure_krb5_conf(cli_realm, cli_domain, cli_server, cli_kdc, dnsok,
```
Comment (stlaz) [inline ipalib/install/host.py]: You can use configure_krb5_conf from the client now ;)

File: ipalib/install/host.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        except OSError:
+            pass
+        try:
+            os.remove(krb_name + ".ipabkp")
+        except OSError:
+            root_logger.error("Could not remove %s.ipabkp", krb_name)
+
+    # DO NOT backup krb5.conf for now
+    # fstore.backup_file(paths.KRB5_CONF)
+    if configure_krb5_conf(
```
Comment (stlaz) [inline ipalib/install/host.py]: bool(configure_krb5_conf) is always false

File: ipaserver/install/server/replicainstall.py [hunk 3]
```diff
@@ -498,6 +499,59 @@ def check_remote_version(api):
...
+    """
+    Do partial host enrollment in DL0:
+        * add host entry to remote master
+        * request temporary keytab from remote master
+        * enroll replica using temporary keytab so that krb5.conf gets
+          configured correctly
+        * add host entry to ipaservers group to elevate its privileges
+    """
+    env = {"PATH": ("/bin:/sbin:/usr/kerberos/bin:"
+                    "/usr/kerberos/sbin:/usr/bin:/usr/sbin")}
```
Comment (stlaz) [inline ipaserver/install/server/replicainstall.py]: This is fugly and should perhaps be wrapped in some platform-dependent module but I realize that's definitely out of our scope at the current moment so I am just ranting here.

**Comment by martbab** [comment]
I have  made substantial changes: I have dropped the host.enroll altogether and user combination of host-add and ipa-getkeytab to enroll the host.

I have also kept hostgroup_add_member in the enrollment function for now. I tried to use the same logic as in promote_check/promote but I came to conclusion that this warrants separate commits during the unification of the install/enroll logic.

File: client/ipa-getkeytab.c [hunk 4]
```diff
@@ -152,10 +154,9 @@ static int ipa_ldap_init(LDAP ** ld, const char * scheme, const char * servernam
 	return rc;
 }
 
-const char *ca_cert_file = "/etc/ipa/ca.crt";
-
 static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
-			 const char *bind_dn, const char *bind_pw, LDAP **_ld)
+			 const char *bind_dn, const char *bind_pw,
+             const char *ca_cert_file, LDAP **_ld)
```
Comment (stlaz) [inline client/ipa-getkeytab.c]: NITPICK_MODE=1
Could you please fix the indentation here?

File: client/man/ipa-getkeytab.1 [hunk 5]
```diff
@@ -21,7 +21,7 @@
 .SH "NAME"
 ipa\-getkeytab \- Get a keytab for a Kerberos principal
 .SH "SYNOPSIS"
-ipa\-getkeytab \fB\-p\fR \fIprincipal\-name\fR \fB\-k\fR \fIkeytab\-file\fR [ \fB\-e\fR \fIencryption\-types\fR ] [ \fB\-s\fR \fIipaserver\fR ] [ \fB\-q\fR ] [ \fB\-D\fR|\fB\-\-binddn\fR \fIBINDDN\fR ] [ \fB\-w|\-\-bindpw\fR ] [ \fB\-P\fR|\fB\-\-password\fR \fIPASSWORD\fR ] [ \fB\-r\fR ]
+ipa\-getkeytab \fB\-p\fR \fIprincipal\-name\fR \fB\-k\fR \fIkeytab\-file\fR[\fB\-e\fR \fIencryption\-types\fR ] [ \fB\-s\fR \fIipaserver\fR ] [\fB\-q\fR] [ \fB\-D\fR|\fB\-\-binddn\fR \fIBINDDN\fR ] [ \fB\-w|\-\-bindpw\fR] [\fB\-P\fR|\fB\-\-password\fR \fIPASSWORD\fR ] [\fB\-c|\-\-cacert \fICACERT\fR][ \fB\-r\fR ]
```
Comment (stlaz) [inline client/man/ipa-getkeytab.1]: Please, leave the spaces around '[' in _encryption-types_

File: ipaserver/install/server/replicainstall.py [hunk 6]
```diff
@@ -32,6 +32,7 @@
     network_ip_address_warning,
     broadcast_ip_address_warning,
 )
+import ipaclient.install.client
```
Comment (stlaz) [inline ipaserver/install/server/replicainstall.py]: I'd rather see `from ipaclient.install import client` or `from ipaclient.install.client import ...`
Is there a reason you're doing the import like this?

File: client/ipa-getkeytab.c [hunk 7]
```diff
@@ -343,6 +344,7 @@ static int ldap_set_keytab(krb5_context krbctx,
 			   krb5_principal princ,
 			   const char *binddn,
 			   const char *bindpw,
+               const char *ca_cert_file,
```
Comment (stlaz) [inline client/ipa-getkeytab.c]: Fix indentation here as well, please.

**Comment by stlaz** [comment]
This makes ipa-replica-install fail on DL0:
http://pastebin.com/hUU6j3E8

**Comment by martbab** [comment]
Then I should prepare a patch that replaces all tabs with spaces in ipa-getkeytab.c, as that is the real issue there (mixed tabs w/ spaces)

**Comment by stlaz** [comment]
ACK, works fine and can be pushed.

File: client/ipa-getkeytab.c [hunk 8]
```diff
@@ -712,6 +717,8 @@ int main(int argc, const char *argv[])
               _("LDAP DN"), _("DN to bind as if not using kerberos") },
 	    { "bindpw", 'w', POPT_ARG_STRING, &bindpw, 0,
               _("LDAP password"), _("password to use if not using kerberos") },
+	    { "cacert", 'c', POPT_ARG_STRING, &ca_cert_file, 0,
```
Comment (jcholast) [inline client/ipa-getkeytab.c]: No short option please, this is not something used frequently.

File: ipaserver/install/server/replicainstall.py [hunk 9]
```diff
@@ -498,10 +499,63 @@ def check_remote_version(api):
...
 
+def enroll_dl0_replica(installer, fstore, remote_api, debug=False):
+    """
+    Do partial host enrollment in DL0:
+        * add host entry to remote master
+        * request temporary keytab from remote master
+        * enroll replica using temporary keytab so that krb5.conf gets
+          configured correctly
+        * add host entry to ipaservers group to elevate its privileges
+    """
```
Comment (stlaz) [inline ipaserver/install/server/replicainstall.py]: Please update the docstring.

**Comment by stlaz** [comment]
Looks fine, hopefully I did not overlook any detail this time, ACK.

**Comment by jcholast** [comment]
Pushed to installer-refactoring.

**Pull request closed without merging**